File: content/LLD/Reference-Code/19. Composite Pattern/bad_example.py
from typing import List


class File:
    def __init__(self, name: str):
        self.__name = name

    def show_details(self):
        return f"File : {self.__name}"


class Folder:
    def __init__(self, name: str):
        self.__name = name
        self.__files: List[File] = []

    def add_file(self, file: File):
        self.__files.append(file)

    def show_details(self):
        print(f"Folder name: {self.__name}")
        for file in self.__files:
            print(file.show_details())


file1 = File("image.png")
file2 = File("ppt")
file3 = File("word.exe")

folder = Folder("my_drive")

folder.add_file(file1)
folder.add_file(file2)
folder.add_file(file3)

folder.show_details()

File: content/LLD/Reference-Code/19. Composite Pattern/BadExample.java
import java.util.ArrayList;
import java.util.List;

public class BadExample {
    static class File {
        private final String name;

        public File(String name) {
            this.name = name;
        }

        public String showDetails() {
            return "File : " + this.name;
        }
    }

    // Flawed: Folder cannot contain subfolders, only files
    static class Folder {
        private final String name;
        private final List<File> files = new ArrayList<>();

        public Folder(String name) {
            this.name = name;
        }

        public void addFile(File file) {
            this.files.add(file);
        }

        public void showDetails() {
            System.out.println("Folder name: " + this.name);
            for (File file : this.files) {
                System.out.println(file.showDetails());
            }
        }
    }

    public static void main(String[] args) {
        File file1 = new File("image.png");
        File file2 = new File("ppt");
        File file3 = new File("word.exe");

        Folder folder = new Folder("my_drive");
        folder.addFile(file1);
        folder.addFile(file2);
        folder.addFile(file3);

        folder.showDetails();
    }
}
